rankanizer/polkadart
# Basic support for System RPC API

Reference:
- https://polkadot.js.org/docs/substrate/rpc#system
- https://github.com/paritytech/substrate/blob/polkadot-v0.9.42/client/rpc-api/src/system/mod.rs#L31-L123
- https://github.com/paritytech/substrate/blob/polkadot-v0.9.42/utils/frame/rpc/system/src/lib.rs#L39-L53

The RPC methods `system_dryRun` and `system_accountNextIndex` are required for sending signed transactions, the usage of `system_accountNextIndex` is recommended over `system.accounts` state, because it also take into account pending transactions and works regardless the storage format.

## RPC methods supported
- [X] `system_name` 
- [X] `system_version`
- [X] `system_chain`
- [X] `system_chainType`
- [X] `system_health`
- [X] `system_localPeerId`
- [X] `system_localListenAddresses`
- [X] `system_peers`
- [x] `system_accountNextIndex`
- [x] `system_dryRun`
- [ ] `system_unstable_networkState` (this one is unstable, we are not going to support it for now).
- [x] `system_addReservedPeer`
- [x] `system_removeReservedPeer`
- [x] `system_reservedPeers`
- [x] `system_nodeRoles`
- [x] `system_syncState`
- [x] `system_addLogFilter`
- [x] `system_resetLogFilter`

## How to test
Add this at the end of the file `packages/polkadart_cli/pubspec.yaml`:
```yaml
polkadart:
  output_dir: bin/generated
  chains:
    polkadot: wss://rpc.polkadot.io
```
then run:
```bash
fvm dart run melos exec --scope="polkadart_cli" -- 'fvm dart run ./bin/generate.dart'
```
Check that Polkadot client now contains `SystemApi` at `packages/polkadart_cli/bin/generated/polkadot/polkadot.dart`

usage:
```dart
import 'package:ss58/ss58.dart' show Address;
import './generated/polkadot/polkadot.dart' show Polkadot;

void main(List<String> arguments) async {
  final Polkadot polkadot = Polkadot.url(Uri.parse('wss://rpc.polkadot.io'));

  // Retrieve node name
  final name = await polkadot.rpc.system.name();
  print('Node name: $name');

  await polkadot.disconnect();
}
```
